Firestore/core/src/firebase/firestore/model/mutation.h

```diff
@@ -199,149 +214,82 @@ class Mutation {
    *     only if maybe_doc was nullopt and the mutation would not create a new
    *     document.
    */
-  virtual absl::optional<MaybeDocument> ApplyToLocalView(
+  absl::optional<MaybeDocument> ApplyToLocalView(
       const absl::optional<MaybeDocument>& maybe_doc,
       const absl::optional<MaybeDocument>& base_doc,
-      const Timestamp& local_write_time) const = 0;
+      const Timestamp& local_write_time) const {
+    return rep_->ApplyToLocalView(maybe_doc, base_doc, local_write_time);
+  }
 
   friend bool operator==(const Mutation& lhs, const Mutation& rhs);
 
- protected:
-  Mutation(DocumentKey&& key, Precondition&& precondition);
+  size_t Hash() const {
+    return rep_->Hash();
```

Comment:
Hrm...  Is it true that dereferencing nullptr is undefined behavior? https://stackoverflow.com/a/28573259

If so and we can't statically prevent calling methods on invalid instances, it seems dangerous not to have explicit checks.  It seems entirely plausible for a coding mistake to result in accidentally calling a method on an invalid instance (e.g. because it was moved-from in another part of the code).

I hoped dereferencing a null shared_ptr<> would fail in a deterministic sane way, but based on https://stackoverflow.com/questions/34409299/why-doesnt-stdshared-ptr-dereference-throw-a-null-pointer-exception-or-simil it sounds like it intentionally doesn't. :-/

So... can you convince me that this will fail in a deterministic, sane way with no possibility of memory corruption bugs?

Reply:
These posts are both problematic because they intentionally steer toward what the language guarantees, and the language gives itself a huge out with the notion of undefined behavior. While it's true that the standard doesn't specify what happens, there are no practically useful systems in the wild where dereferencing null doesn't crash somehow. On UNIX systems including the iPhone, the process will receive signal 11 (SEGV) and that's the end of it.

Admittedly this may not result in a stack trace. Under Xcode it certainly pauses the process exactly at the point of failure, but this isn't going to happen for end users. A hard assert is slight better in that it will result in a stack.

Note that the code this is replacing wasn't any better. Using moved-from `std::shared_ptr<Mutation>` is equivalently hazardous, though perhaps by making the pointer more obvious invites more caution. IDK.

I'm willing to insert the checks if we can do so without littering it all over the code. The downside is that if it's in a central location it may be harder to pinpoint what's going on.

Anyway, to sum up:
  * Users will definitely crash on a null pointer dereference
  * It's not likely to be all that useful a message
  * Centralizing the check would make it not too cumbersome
  * I worry that the result may not end up all that useful either.

Reply:
I think ensuring we get a stack trace is probably worthwhile.  To centralize we could probably put the check in the `rep()` getter and then change all the implementation methods to go through the getter.

Reply:
I've reimplemented `rep()` as `*NOT_NULL(rep_)`.

Also since we're going to be null checking a lot more with this, I made null checks cheaper (by ~30%).